// ngx-modal-dialog/index.js
export * from './src/simple-modal.component';
export * from './src/modal-dialog.service';
export * from './src/modal-dialog.component';
// components and directives
import { ModalDialogComponent } from './src/modal-dialog.component';
import { ModalDialogService } from './src/modal-dialog.service';
import { SimpleModalComponent } from './src/simple-modal.component';
import { ModalDialogInstanceService } from './src/modal-dialog-instance.service';
// modules
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
var ModalDialogModule = (function () {
    function ModalDialogModule() {
    }
    ModalDialogModule.forRoot = function () {
        return {
            ngModule: ModalDialogModule,
            providers: [ModalDialogInstanceService]
        };
    };
    return ModalDialogModule;
}());
export { ModalDialogModule };
ModalDialogModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                declarations: [ModalDialogComponent, SimpleModalComponent],
                entryComponents: [ModalDialogComponent, SimpleModalComponent],
                exports: [ModalDialogComponent, SimpleModalComponent],
                providers: [ModalDialogService]
            },] },
];
/** @nocollapse */
ModalDialogModule.ctorParameters = function () { return []; };
